### PR TITLE
Avoid buffered I/O

### PR DESCRIPTION
**Pull request by RinHizakura**
Here's a macro solution to avoid unexpected output by the buffered I/O `printf`. It could have some limitations and need improvement, but it achieves the purpose of output in text lines as we may want.

### CODE REVIEW

File: fiber/fiber.c [hunk 1]
```diff
@@ -119,13 +119,24 @@ int fiber_wait_all()
     return FIBER_NOERROR;
 }
 
+/* A simple solution to avoid unexpected output by the buffered I/O printf is
+ * using low-level I/O interface. This is a naive implemention for the idea, and
+ * it will fail if the formatted string with length >64 */
+#define SAFE_PRINT(fmt, ...)                                  \
+    do {                                                      \
+        char str[64];                                         \
```
Comment (jserv) [inline fiber/fiber.c]: Use macro to specify the static storage length for `str`.

File: fiber/fiber.c [hunk 2]
```diff
@@ -119,13 +119,24 @@ int fiber_wait_all()
     return FIBER_NOERROR;
 }
 
+/* A simple solution to avoid unexpected output by the buffered I/O printf is
```
Comment (jserv) [inline fiber/fiber.c]: You shall explain why this wrapper is "safe."

File: fiber/fiber.c [hunk 3]
```diff
@@ -119,13 +119,24 @@ int fiber_wait_all()
...
+ * using low-level I/O interface. This is a naive implemention for the idea, and
+ * it will fail if the formatted string with length >64 */
+#define SAFE_PRINT(fmt, ...)                                  \
+    do {                                                      \
+        char str[64];                                         \
+        int n = sprintf(str, fmt __VA_OPT__(, ) __VA_ARGS__); \
+        write(STDOUT_FILENO, str, n);                         \
+    } while (0)
+
+
```
Comment (jserv) [inline fiber/fiber.c]: Remove one blank line.

File: fiber/fiber.c [hunk 4]
```diff
@@ -119,13 +119,24 @@ int fiber_wait_all()
     return FIBER_NOERROR;
 }
 
+/* A simple solution to avoid unexpected output by the buffered I/O printf is
+ * using low-level I/O interface. This is a naive implemention for the idea, and
+ * it will fail if the formatted string with length >64 */
+#define SAFE_PRINT(fmt, ...)                                  \
```
Comment (jserv) [inline fiber/fiber.c]: Rename to `printf_safe` to emphasize the variant.

File: fiber/fiber.c (hunk 2 above)
Comment (RinHizakura) [reply fiber/fiber.c]: Actually, I am thinking to change the name for the macro.  I use the word "safe" to represent that the output can show as we may want. But as far as I know, `printf` is said to be thread-safe in the Linux manual. That means it won't corrupt anything when using in the multithreading. However, some interleaving that doesn't change the outcome is still allowed to happen, which can produce the mixing output between threads. 

So I am not sure if the name `print_safe` will mislead someone to the concept that `printf` is not thread-safe. Do I misunderstand something or any advice for this?

Comment (jserv) [reply fiber/fiber.c]: It was meant to be "safe." Instead, its intention is to eliminate the impact of buffered I/O, which causes the unexpected sequence of strings.

Comment (jserv) [reply fiber/fiber.c]: While the `printf` provided by glibc is a buffered output function, the proposed change does "direct" format string manipulation. Therefore, you can rename `printf_safe` to `printf_unbuffered` and then use `#define printf printf_unbuffered` for build-time substitution.

Reference: [CS:APP Chapter 10](http://csapp.cs.cmu.edu/2e/ch10-preview.pdf)

File: fiber/fiber.c [hunk 5]
```diff
@@ -19,6 +19,26 @@
 #include <sys/wait.h>  /* For wait */
 #include <unistd.h>    /* For getpid */
 
+/* A simple solution to avoid unexpected output by the buffered I/O 'printf' is
+ * using low-level I/O interface 'write'. It works because 'printf' will wait to
+ * write STDOUT until the buffer is full, or on some other conditions. Using
+ * write, however, can write STDOUT immediately.
```
Comment (jserv) [inline fiber/fiber.c]: Change "write STDOUT" to "write to STDOUT"

**Comment by RinHizakura** [comment]
When I try to go deep on my solution for the interleaved output, I think I may have some wrong thoughts when I design this macro. I first used this as the solution because I think `write` won't be interrupted except the signal handler during the execution, so it can output the whole given string immediately without interleaving.  But soon after some exploration and thinking, I doubt if I have to handle the possibility that `write` may write less than the request `n` bytes.

Although for my result, the output contents do show as expected, but I'm not sure if this really solves the problem. I may need time to clarify my own question.

**Pull request merged**